### DISCUSSION TITLE
Has anybody ever used LmHandler in their own project successfully?

### DISCUSSION THREAD

**Discussion by Cjtechnologyltd** [Lora-net/LoRaMac-node#1380] [Q&A]
Hello, I am currently using the latest of master (26dd83c2), ported to a Maxim Cortex M4 in EU868 region with sx1262 radio. I sucessfully ported spi, gpio and rtc and have confirmed that they work as expected.

Following the example laid out in "periodic uplink lpp" for the stm32l476, I had tried to implement my own version where I send up-links on receiving a uart command.

The issue I'm facing is non-repeatability, ie sometimes it "just works", does a join and I can send packets. Other times, I do a request a join, MLME request callback says OK and then nothing else happens...and then the stack is supposedly busy doing.....something and never leaves the busy state. Other times, we do a join but the Rx1 window event just doesn't happen and I miss the join accept.

Also, oddly, when using printf statements at certain places in LoraMac.c, this also seems to break the code. (Places such as at the bottom of OnRxWindow1TimerEvent)

**Comment by zouOrange** [comment]
I am using loraWAN1.0.3, I have used LmHandler, and I wrote AT instructions to complete Class A and Class C.

**Reply by Cjtechnologyltd** [reply]
Please may I ask what hardware(micro and radio) you are using and if you had any issues with the porting stage?

**Reply by zouOrange** [reply]
I am using sx1262 RF. Of course there were some problems, but they were minor. Patience debugging.

**Comment by dalbert2** [comment]
I am also using LoRaWAN 1.0.3 on an STM32L4 uC with an SX1276 xcvr.
I started with the stack provided as part of the STM32Cube LoRaWAN extension.  Unfortunately it had to be integrated into an existing product CMSIS-based codebase (not developed using HAL or the STM32Cube ecosystem); it was a bear.  However, once successfully ported, it seems quite reliable.  As @zouOrange said, patience debugging.

It would be immensely helpful if there were more documentation available.

**Reply by Cjtechnologyltd** [reply]
I got it working eventually, main issues were the SPI and RTC ports where I needed to heavily optimise.

Some documentation would be useful, like more than nothing otherwise I fear this may become abandon-ware when a competitor arrives.

I'm having similar issues with MACv5 and regular LoRa..no documentation of the driver at all, examples don't work.

Just praying for a competitor to Semtech.

**Reply by Cjtechnologyltd** [reply]
Although it was still "off"and intermittent on whether we'd receive a packet or not. Most issues seem to be related to the radio peripheral itself.

**Reply by dalbert2** [reply]
Glad you got it working!  I agree that the lack of adequate documentation, particularly a how-to for porting the stack to a new platform and compliance testing is conspicuously absent and a serious obstacle to adoption.

**Reply by jamesl-dm** [reply]
> Just praying for a competitor to Semtech.

NB-IoT chipsets are sub $5 now :)
